sdk/search/azure-search-index/design.md

```diff
@@ -0,0 +1,483 @@
+
+client = SearchIndexClient(
+    search_service_name="my-search-service",
+    index_name="hotels-sample-index",
+    credential=SearchApiKeyCredential(api_key="xxxxxxxxx")
+)
+
+result = client.get_document(key="10")
+
+print(result)
+```
+
+### 3. Search for documents in an Azure search index with simple text
+
+```python
+from azure.search import SearchApiKeyCredential, SearchIndexClient
+
+client = SearchIndexClient(
+    search_service_name="my-search-service",
+    index_name="hotels-sample-index",
+    credential=SearchApiKeyCredential(api_key="xxxxxxxxx")
+)
+
+results = client.get_search_results(search_text="WiFi")
+
+for doc in results:
+    print(doc['Description'])
+    print(doc['Type'])
+    print(doc['BaseRate'])
+```
+
+### 3. Filter results Azure search index
+
+```python
+from azure.search import SearchApiKeyCredential, SearchIndexClient, SearchQuery
+
+client = SearchIndexClient(
+    search_service_name="my-search-service",
+    index_name="hotels-sample-index",
+    credential=SearchApiKeyCredential(api_key="xxxxxxxxx")
+)
+
+query = SearchQuery(
+    search_text="WiFi",
+    highlightPreTag="<em>",
+    highlightPostTag="</em>")
+query.fiter(AND("Address/City eq 'Portland'",  "Address/Country eq 'US'"))
+query.orderby(desc("Rating"), asc(geo_distance("location", POINT(10, 20))))
+query.select("HotelId", "HotelName", "Rating")
+
+results = client.get_search_results(search_query=query)
+
+for doc in results:
+    print(doc['HotelName'])
+    print(doc['Rating'])
+
+```
+
+### 5. Get a list of search suggestions
+
+```python
+from azure.search import SearchApiKeyCredential, SearchIndexClient
+
+client = SearchIndexClient(
+    search_service_name="my-search-service",
+    index_name="hotels-sample-index",
+    credential=SearchApiKeyCredential(api_key="xxxxxxxxx")
+)
+
+results = client.get_suggestions(search_text="cof", suggester_name="sg")
+
+for result in results:
+    print(result['text'])
+```
+
+### 6. Get a list of search auto-completions
+
+```python
+from azure.search import SearchApiKeyCredential, SearchIndexClient
+
+client = SearchIndexClient(
+    search_service_name="my-search-service",
+    index_name="hotels-sample-index",
+    credential=SearchApiKeyCredential(api_key="xxxxxxxxx")
+)
+
+results = client.get_autocompletions(search_text="cof", suggester_name="sg")
+
+for result in results:
+    print(result['text'])
+    print(result['query_plus_text'])
+```
+
+### 7. Upload documents to an Azure Search index
+
+```python
+from azure.search import SearchApiKeyCredential, SearchIndexClient
+
+client = SearchIndexClient(
+    search_service_name="my-search-service",
+    index_name="hotels-sample-index",
+    credential=SearchApiKeyCredential(api_key="xxxxxxxxx")
+)
+
+documents = [
+    { "id": "10", "description": "new item" },
+    { "id": "11", "description": "another new item" },
+]
+
+results = client.upload_documents(documents)
+
+for result in results:
+    print(result.status_code)
```

Comment:
Let's see if we can uplevel this somewhat - for most services, we've tried to remove the need to explicitly check for http status codes. For text analytics, we are adding an `is_error` attribute on the response element (and in that case, there is no per-response status code). It would be great to unify the pattern here.

Reply:
is there a PR I can refer to to keep consistency?

Reply:
Not so much a PR, but you can look [here](https://github.com/Azure/azure-sdk-for-python/blob/master/sdk/textanalytics/azure-ai-textanalytics/azure/ai/textanalytics/_models.py#L108). Any result object will have `is_error=False` and the error obj will have `is_error=True`

Reply:
Assuming we’re talking about the per-document status code returned by the Index API, a Boolean flag isn’t going to be enough (there is already one of those too). The status code is used to determine how to recover from failure to index a particular document, and whether the error is retryable. The codes are documented here: https://docs.microsoft.com/rest/api/searchservice/addupdate-or-delete-documents#response

In the Track 1 .NET SDK, we took the approach of providing a default implementation of the logic that checks per-document status codes. The flow goes like this:
  1. Call Index, passing a batch of actions. 
  1. Partial failure occurs; The REST API returns 207 and a list of results, one per document, where some of them have failure status codes. Our client library takes the deserialized results, wraps them in an IndexBatchException, and throws it to the caller (the rationale for using exceptions for this case is that we don’t want customers to ignore partial failures by default).
  1. Customer code catches IndexBatchException and calls its FindFailedActionsToRetry method, passing the original batch. It returns a filtered version of that batch that includes only the failed actions that are retryable (e.g. 404 is not retryable). 

The code for FindFailedActionsToRetry is here: https://github.com/Azure/azure-sdk-for-net/blob/bd0af1d51eaaaf8f169a3b437d1a66447e30d7a4/sdk/search/Microsoft.Azure.Search.Data/src/Customizations/IndexBatchException.cs#L49

If you still want to up-level the representation of the status code in the results object, I would suggest an object with the Python equivalent of a meaningfully-named enum, as well as the integer status code alongside it (so customers can look it up in the REST API docs if needed).

Reply:
We generally don't expose the status code in our Python libraries other than for failures (and then through exceptions). I would consider automatically retrying on things that are known to be retry:able (up to the number of retries the client requested) if that is what we tell our customers to do. 

How frequently are non-retryable errors occurring? And is there any categorization that can be done (e.g. precondition failed/not found etc.)?